### DISCUSSION TITLE
Question about blur system

### DISCUSSION THREAD

**Discussion by Dickby** [hyprwm/Hyprland#3820] [General]
Hi,
i'm looking at the way the blur is done at the moment, because i think it can be optimized in some ways.
I have a hard time to understand, why the vibrancy code is inside the blur1 shader.
It's totally possible that this makes sense, and i just don't get it, but why should the vibrancy be applied for every downscaling blur step? Shouldn't it just be applied once after blurring?

**Comment by vaxerski** [comment]
Nope. @Visual-Dawg might have a better explanation than me though.

**Comment by vdawg-git** [comment]
The original idea was to let strong colors "spread/grow" with the blur.
![image](https://github.com/hyprwm/Hyprland/assets/28539403/dfd75c55-4963-4ee9-a3b5-1e01afee5f4d)

With my understanding, this requires an increase in saturation for saturated colors between each pass, so that they can grow into neutral colors.
However, I wasnt able to recreate that just yet.

-----
Base
![image](https://github.com/hyprwm/Hyprland/assets/28539403/06686cd3-f2de-4c07-9ace-a3deeca320a4)
-----
Saturation applied after blur
- Creates banding
- (Nearly) complimentary create a gray in-between gradient
![image](https://github.com/hyprwm/Hyprland/assets/28539403/78ab3f59-acf3-4ea6-a862-9fe1d253ecc1)
----
Saturation applied between passes
- No banding
- Smooth color gradient.
![image](https://github.com/hyprwm/Hyprland/assets/28539403/f7f6e7c1-9e84-4bfd-b99c-1db9405ebbbd)
-----
Ideal result for spreading color
- Very nice strong purple between blue and orange
- Strong bright colors spread outwards (most noticeably the green here)
![image](https://github.com/hyprwm/Hyprland/assets/28539403/0ce2ad99-f3fc-43bf-9f74-819a2eefb70c)

I achieved this effect by blurring, then increasing the saturation and decreasing the brightness minimally (-1 in Photopea). Then repeating this process.


If you find a way to make the "spreading" colors work with HSL, this would be really cool :)

Edit: Feel free to join the Discord and suggest/discuss better ways there too :D

**Comment by Dickby** [comment]
@Visual-Dawg: Thanks for the detailed explanation. 
It  seems a bit expensive, but maybe it's needed for the desired effect, i will think about it.
At least with `vibrance` as a uniform variable, it should have no effect on performance if vibrance is disabled.
Btw. with your explanation, shouldn't saturation be applied while upscaling too?

**Reply by vdawg-git** [reply]
Yes, if set to 0 it shouldnt cost anything. 

Why should it be applied while upscaling too?

And it would be really cool if you can help with the shader :D

**Reply by Dickby** [reply]
> Why should it be applied while upscaling too?

The Dual Kawase is using the down and upscaling passes to gather and mix the colors from the surrounding pixels.
So my naive question is, why should saturation only be applied by half of the gathering/mixing operations?

**Reply by vdawg-git** [reply]
Good question. I also do not fully understand the process. 
Would it make a huge difference if it is also applied during the upscaling?

Because I think that most of the blurring occurs during the upscaling--..
Hmm, you might have a point here - 

But vibrancy gets applied *after* the downscaling, thus before the upscaling. 
The values mix the most during upscaling (I think, but I might be wrong), and it is a lot faster to apply the effect on the downscaled version.

Your thoughts?

**Reply by Dickby** [reply]
Dual Kawase is emulating a Gaussian blur by this scaling operations, the performance benefit compared to a gaussian kernel comes from:
1. operating on smaller textures due to the downscaling
2. both the down- and upscaling target pixels in between, so the graphics card mixes those colors automatically. This way you get the value of 2 to 4 neighbouring pixels from just one "in-between-pixel".

So the blurring is really done by both operations.
The last upscaling pass should be the most expensive, because it's the only pass that operates on a full size texture, i think that is the reason why the saturation in upscaling pass has a bigger effect on performance. 

The optimizations i have in mind are not related to this though, i was just wondering saturation was applied this way.
I will think a bit about the math of saturation combined with the Dual Kawase to achieve that color spreading too.
It's an interesting problem and I agree that it's a nice effect.
